baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreferInputStreamTransferTo.java

```diff
@@ -0,0 +1,115 @@
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8067661">https://bugs.openjdk.org/browse/JDK-8067661</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8265891">https://bugs.openjdk.org/browse/JDK-8265891</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8273038">https://bugs.openjdk.org/browse/JDK-8273038</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8279283">https://bugs.openjdk.org/browse/JDK-8279283</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8296431">https://bugs.openjdk.org/browse/JDK-8296431</a></li>
+ *  </ul>
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = SeverityLevel.WARNING,
+        summary = "Prefer JDK `InputStream.transferTo(OutputStream)` over utility methods such as "
+                + "`com.google.common.io.ByteStreams.copy(InputStream, OutputStream)`, "
+                + "`org.apache.commons.io.IOUtils.copy(InputStream, OutputStream)`, "
+                + "`org.apache.commons.io.IOUtils.copyLong(InputStream, OutputStream)`, "
+                + "cf. https://github.com/palantir/gradle-baseline/issues/2615")
+public final class PreferInputStreamTransferTo extends BugChecker implements BugChecker.MethodInvocationTreeMatcher {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final String ERROR_MESSAGE = "Prefer InputStream.transferTo(OutputStream)";
+
+    public static final String INPUT_STREAM = "java.io.InputStream";
+    private static final Matcher<Tree> INPUT_STREAM_MATCHER = MoreMatchers.isSubtypeOf(INPUT_STREAM);
+    public static final String OUTPUT_STREAM = "java.io.OutputStream";
+    private static final Matcher<Tree> OUTPUT_STREAM_MATCHER = MoreMatchers.isSubtypeOf(OUTPUT_STREAM);
+
+    private static final Matcher<ExpressionTree> GUAVA_BYTE_STREAM_COPY_MATCHER = MethodMatchers.staticMethod()
+            .onDescendantOf("com.google.common.io.ByteStreams")
+            .namedAnyOf("copy")
+            .withParameters(INPUT_STREAM, OUTPUT_STREAM);
+    private static final Matcher<ExpressionTree> APACHE_COMMONS_BYTE_STREAM_COPY_MATCHER = MethodMatchers.staticMethod()
+            .onDescendantOf("org.apache.commons.io.IOUtils")
+            .namedAnyOf("copy", "copyLarge")
+            .withParameters(INPUT_STREAM, OUTPUT_STREAM);
+    private static final Matcher<ExpressionTree> AWS_BYTE_STREAM_COPY_MATCHER = MethodMatchers.staticMethod()
+            .onDescendantOf("com.amazonaws.util.IOUtils")
+            .namedAnyOf("copy")
+            .withParameters(INPUT_STREAM, OUTPUT_STREAM);
+
+    private static final Matcher<ExpressionTree> METHOD_MATCHER = Matchers.anyOf(
+            GUAVA_BYTE_STREAM_COPY_MATCHER, APACHE_COMMONS_BYTE_STREAM_COPY_MATCHER, AWS_BYTE_STREAM_COPY_MATCHER);
```

Comment:
I started with these mostly for simplicity and to cover the common usages, and may expand in follow on PR. Hadoop [`IOUtils.copyBytes`](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/io/IOUtils.html#copyBytes-java.io.InputStream-java.io.OutputStream-org.apache.hadoop.conf.Configuration-) methods are some the worst offenders I've seen as they default to 4KiB chunks 😢